### ISSUE TITLE
Blockscout doesn't see minting ERC20 tokens because of system call

### ISSUE THREAD

**Issue by varasev** [poanetwork/blockscout#2168]
Blockscout doesn't update ERC20 contract balance (Total Supply) and its token holder ERC20 balances while the ERC20 tokens minting occurs inside a system call. This happens when trying to launch Blockscout locally along with the [`posdao-test-setup`](https://github.com/poanetwork/posdao-test-setup).

Example:

When `BlockReward.reward` function is [called by the system](https://github.com/poanetwork/posdao-contracts/blob/99da89cd87067aa33471ad1931e333babd5d7e9a/contracts/BlockRewardAuRa.sol#L21), it calls the [`_distributeRewards` internal function](https://github.com/poanetwork/posdao-contracts/blob/99da89cd87067aa33471ad1931e333babd5d7e9a/contracts/BlockRewardAuRa.sol#L113) which in turn calls the [`mintReward` function](https://github.com/poanetwork/posdao-test-setup/blob/5b1bc9fa58432fb802cfbb2c662d70ba673bfe6b/contracts/ERC677BridgeTokenRewardableMock.sol#L598) of the ERC20 contract: https://github.com/poanetwork/posdao-contracts/blob/99da89cd87067aa33471ad1931e333babd5d7e9a/contracts/BlockRewardAuRa.sol#L372

When that occurs, the `Total Supply` value on the page http://localhost:4000/tokens/0x6f7a73c96bd56f8b0debc795511eda135e105ea3/token_transfers must be increased every 80 blocks and the list of the token holders must be actual (with their actual balances): http://localhost:4000/tokens/0x6f7a73c96bd56f8b0debc795511eda135e105ea3/token_holders

Steps to reproduce:

`1.` Launch `posdao-test-setup`.

`1.1.` Perform the commands:

```bash
$ git clone -b aura-pos https://github.com/poanetwork/parity-ethereum
$ cd parity-ethereum
$ cargo build --release --features final
$ cd ..
$ git clone https://github.com/poanetwork/posdao-test-setup
$ cd posdao-test-setup
```

`1.2.` Add the options

```
interface = "all"
origins = ["all"]
```

to the section `[websockets]` in the `config/node1.toml`.

`1.3.` Perform `npm run all` command.

`2.` Launch the Blockscout in a separate console with docker (for Linux use `localhost` instead of `host.docker.internal`):

```bash
$ git clone -b staking-web-app https://github.com/poanetwork/blockscout
$ cd blockscout/docker
$ export ETHEREUM_JSONRPC_HTTP_URL=http://host.docker.internal:8541 # use localhost here for Linux
$ export ETHEREUM_JSONRPC_TRACE_URL=http://host.docker.internal:8541 # use localhost here for Linux
$ export ETHEREUM_JSONRPC_WS_URL=ws://host.docker.internal:9541 # use localhost here for Linux
$ export POS_VALIDATORS_CONTRACT=0x1000000000000000000000000000000000000001
$ export POS_STAKING_CONTRACT=0x1100000000000000000000000000000000000001 
$ export ETHEREUM_JSONRPC_VARIANT=parity
$ export COIN=POA
$ export NETWORK=POA
$ export LOGO=/images/posdao_logo.png
$ export LOGO_FOOTER=/images/posdao_logo_footer.png
$ export MIN_CANDIDATE_STAKE=1
$ export MIN_DELEGATOR_STAKE=1
$ make start
```

`3.` Open http://localhost:4000/tokens/0x6f7a73c96bd56f8b0debc795511eda135e105ea3/token_transfers and wait for block `#229` is mined. After that block, the `Total Supply` must be more than 36 tokens but it displays strictly `36 POS`. Each of 10 tokens holders displayed on the tab http://localhost:4000/tokens/0x6f7a73c96bd56f8b0debc795511eda135e105ea3/token_holders must show more than `2 POS` tokens but they show strictly `2 POS` each.

**Comment by varasev** [comment]
Also, note that the [`mintReward`](https://github.com/poanetwork/posdao-test-setup/blob/5b1bc9fa58432fb802cfbb2c662d70ba673bfe6b/contracts/ERC677BridgeTokenRewardableMock.sol#L598) function cannot emit [events](https://github.com/poanetwork/posdao-test-setup/blob/5b1bc9fa58432fb802cfbb2c662d70ba673bfe6b/contracts/ERC677BridgeTokenRewardableMock.sol#L606-L607) because the system call doesn't support emitting events.

**Comment by ayrat555** [comment]
@varasev we have `Token.MetadataUpdater` which periodically make requests to a node to ask for a `totalSupply`

**Comment by varasev** [comment]
> we have `Token.MetadataUpdater` which periodically makes requests to a node to ask for a `totalSupply`

OK. What's the period of making those requests? Is it configurable?

Token holder balances should also be updated somehow.

**Comment by ayrat555** [comment]
@varasev it updates every 2 days. the period is configurable. token balances are updated based on token transfers in transactions

**Comment by varasev** [comment]
> it updates every 2 days. the period is configurable

OK

> token balances are updated based on token transfers in transactions

I.e. it listens for the ERC20 `Transfer` event, right? As I [mentioned above](https://github.com/poanetwork/blockscout/issues/2168#issuecomment-502010012) unfortunately we don't have transactions when minting tokens, so there are no events.

**Comment by saneery** [comment]
Maybe we should set an environment variable to decrease the period of updating

**Comment by vbaranov** [comment]
>it updates every 2 days. the period is configurable.

Yes, metadata_updater_days_interval is hardcoded
https://github.com/poanetwork/blockscout/blob/7e13e38f3abc85a1fbc3b2411235bb268d410081/apps/indexer/config/config.exs#L34

It would be nice to have ENV var for that

**Comment by ayrat555** [comment]
I'll make a PR after my current task. (1-2 hours)

**Comment by ayrat555** [comment]
@varasev we can implement token transfer detection logic specific to your case. what can we use to detect these token transfers?

**Comment by varasev** [comment]
> we can implement token transfer detection logic specific to your case. what can we use to detect these token transfers?

I'll think about it...

**Comment by varasev** [comment]
So, the logic for periodic updating token holder balances assumes to have the following two steps:

1. As soon as we reach the block number
    [`StakingAuRa.stakingEpochStartBlock()`](https://github.com/poanetwork/posdao-contracts/blob/3e5b39676daac153d02ff6a95975acd058f69f6d/contracts/StakingAuRa.sol#L111) + 
[`ValidatorSetAuRa.MAX_VALIDATORS()`](https://github.com/poanetwork/posdao-contracts/blob/3e5b39676daac153d02ff6a95975acd058f69f6d/contracts/abstracts/ValidatorSetBase.sol#L18) * [`BlockRewardAuRa.DELEGATORS_ALIQUOT()`](https://github.com/poanetwork/posdao-contracts/blob/3e5b39676daac153d02ff6a95975acd058f69f6d/contracts/abstracts/BlockRewardBase.sol#L21) + `2`,
    we should make an array of all addresses which ERC20 balances must be updated.

    To do that, we need to call the [`BlockRewardAuRa.snapshotStakingAddresses()`](https://github.com/poanetwork/posdao-contracts/blob/3e5b39676daac153d02ff6a95975acd058f69f6d/contracts/abstracts/BlockRewardBase.sol#L290) to get the snapshotted pool addresses and then for each of these addresses we should call the [`BlockRewardAuRa.snapshotStakers(address)`](https://github.com/poanetwork/posdao-contracts/blob/3e5b39676daac153d02ff6a95975acd058f69f6d/contracts/abstracts/BlockRewardBase.sol#L281) in cycle. This function returns an array of addresses for which we will update ERC20 balances. That way we need to collect all the retrieved addresses in a single array.

    The array of addresses must be saved (in a postgres database or somewhere else) to be used on the step `2`.

2. After we have formed the address array on the step `1`, we should wait until the current staking epoch finishes, i.e. the block number [`StakingAuRa.stakingEpochEndBlock()`](https://github.com/poanetwork/posdao-contracts/blob/3e5b39676daac153d02ff6a95975acd058f69f6d/contracts/StakingAuRa.sol#L116) is reached. After that block is reached, we need to call the `balanceOf` getter of ERC20 contract (which address can be retrieved with [`StakingAuRa.erc20TokenContract()`](https://github.com/poanetwork/posdao-contracts/blob/3e5b39676daac153d02ff6a95975acd058f69f6d/contracts/abstracts/StakingBase.sol#L428) getter) for each of the addresses we saved on step `1` and update their ERC20 balances for that ERC20 contract in the Blockscout database.

We have to repeat these two steps every staking epoch to keep the balances up-to-date.

It's important to note that the process of step 2 must not overlap the process of step 1. I.e. step 2 must be finished before the next step 1 starts.

**Comment by goodsoft** [comment]
@varasev I'm trying to understand this issue right now, in order to solve it efficiently without kludges.
Can you please confirm or disprove the following statements?
1. The problem concerns the phase of _distributing rewards_ that happens once every _epoch_.
2. Reward coins get minted and distributed across validators, which in turn increases token supply.
3. There is neither a transaction, nor an event generated as a result of distributing rewards.
4. The only way to detect these changes is fetching token balance for _every_ possibly affected address, i.e. for all stakers in a particular epoch.

**Comment by varasev** [comment]
Yes, everything is right. As for point 2: The reward coins are distributed across validators and their delegators.

**Comment by goodsoft** [comment]
Fixed in #2471
